backend/fregepoc/indexers/utils/bitbucket.py

```diff
@@ -0,0 +1,144 @@
+    fetch the next one.
+    """
+
+    after_datetime_as_iso = datetime.datetime.isoformat(after_datetime)
+
+    response = requests.get(
+        API_REPOSITORIES_ENDPOINT,
+        params={"pagelen": 1, "after": after_datetime_as_iso},
+    )
+
+    if not response:
+        return None, None
+
+    response_json = response.json()
+
+    repository_data_values = response_json.get("values")
+    repository_data = (
+        repository_data_values[0] if repository_data_values else None
+    )
+
+    next_url = response_json.get("next")
+    new_after_datetime = (
+        _parse_datetime_from_next_url(next_url) if next_url else None
+    ) or datetime.datetime.utcnow()
+
+    return repository_data, new_after_datetime
+
+
+def get_forks_count(repository_data: RepositoryData) -> int:
+    """Retuns the number of forks of given repository or `0` on error."""
+    url = _safe_get(repository_data, ["links", "forks", "href"])
+
+    if not url:
+        return 0
+
+    response = requests.get(url, params={"pagelen": 0})
+
+    if not response:
+        return 0
+
+    forks_count = response.json().get("size", 0)
+
+    return forks_count
+
+
+def get_clone_url(repository_data: RepositoryData) -> str | None:
+    """Retuns the clone URL for a repository if it's hosted on BitBucket."""
+    clone_objects = _safe_get(repository_data, ["links", "clone"]) or []
+
+    for clone_details in clone_objects:
+        clone_type = clone_details.get("name")
+        clone_url = clone_details.get("href")
+
+        # reject repositoriest from github.com etc.
+        clone_url_host = parse.urlparse(clone_url).netloc
+
+        if (
+            clone_type in {"http", "https"}
+            and clone_url_host == BITBUCKET_HOSTNAME
+        ):
+            return clone_url
+
+
+def get_repo_url(repository_data: RepositoryData) -> str | None:
+    """Retuns the URL of a repository's webpage."""
+    return _safe_get(repository_data, ["links", "html", "href"])
+
+
+def get_last_commit_hash(repository_data: RepositoryData) -> str | None:
+    """Retuns the SHA256 hash of the repository's latest commit."""
+    url = _safe_get(repository_data, ["links", "commits", "href"])
+
+    if not url:
+        return None
+
+    response = requests.get(url, params={"pagelen": 1})
+
+    if not response:
+        return None
+
+    commits = response.json().get("values")
+
+    if commits:
+        return commits[0].get("hash")
+
+
+def _parse_datetime_from_next_url(url: str) -> datetime.datetime | None:
+    """
+    Given a URL, parses the datetime included as its querystring parameter.
+
+    The datetime is expected to be present under 'after' parameter and
+    represented in ISO 8601 format.
+    """
+    parsed_url = parse.urlparse(url)
+    query_string = parse.parse_qs(parsed_url.query)
+    after_params = query_string.get("after")
+
+    if after_params:
+        return datetime.datetime.fromisoformat(after_params[0])
+
+
+def _safe_get(root: dict, keys: list) -> Any:
```

Comment:
Isn't there a more precise type than `Any` we can use to annotate this function?

Reply:
It can be literally anything, depending on contents of the `root` dictionary.
Theoretically, I could write `str | None` – but this is only because I know that at this point the function is used only for retrieving strings. But then, the typing would be wrong, because this function is not limited to retrieving `str | None` but any nested value.

Reply:
Ok, I understand. You can leave it as it is.